File: 101.symmetric-tree.py
#
# @lc app=leetcode id=101 lang=python3
#
# [101] Symmetric Tree
#

# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        if not root.left and not root.right:
            return True
        if not root.left or not root.right:
            return False
        
        return self.isMirror(root.left,root.right)

    def isMirror(self,root1:TreeNode,root2:TreeNode):
        if not root1 and not root2:
            return True
        if not root1 or not root2:
            return False
        if not root1.val == root2.val:
            return False
        return self.isMirror(root1.left,root2.right) and self.isMirror(root1.right,root2.left)

        
# @lc code=end

File: 101.symmetric-tree.go
/*
 * @lc app=leetcode id=101 lang=golang
 *
 * [101] Symmetric Tree
 */
package leetcode

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 1. 层顺遍历，然后判断每层是否是回文，实现见
// 1.1 层序遍历见 102 103
// 1.2 回文判断见 125
// 2. 判断左子树和右子树是否镜像
// 边界条件： 空树
func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	return isMirror(root.Left, root.Right)

}
func isMirror(p, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}
	if p == nil || q == nil {
		return false
	}
	if p.Val != q.Val {
		return false
	}
	return isMirror(p.Left, q.Right) && isMirror(p.Right, q.Left)

}

// @lc code=end
